vowpalwabbit/core/src/reductions/cb/details/large_action/spanner_impl.cc

```diff
@@ -8,65 +8,155 @@ namespace VW
 {
 namespace cb_explore_adf
 {
-std::pair<float, uint64_t> spanner_state::find_max_volume(Eigen::MatrixXf& U, uint64_t X_rid, Eigen::MatrixXf& X)
+void find_fast_max_vol(const Eigen::MatrixXf& U, const Eigen::VectorXf& Sinvtopei, float& max_volume, uint64_t& U_rid)
 {
-  // Finds the max volume by replacing row X[X_rid] with some row in U.
-  // Returns the max volume, and the row id of U used for replacing X[X_rid].
-
-  float max_volume = -1.0f;
-  uint64_t U_rid{};
-  const Eigen::RowVectorXf original_row = X.row(X_rid);
-
   for (auto i = 0; i < U.rows(); ++i)
```

Comment:
Here we don't need to check all U rows, since some rows are already used in X, and using a duplicate row will make the determinant zero, and thereby not choosing it, but we'll pay the determinant computation cost. One potential optimization is to pass in the bit vector to eliminate already used rows, instead of trying all U rows, which might be more useful when d is big.

Reply:
I'll benchmark this and see if it should be implemented next